### PR TITLE
[TS migration] Migrate 'WorkspaceInvite' page to TypeScript

### PR DESCRIPTION
**Pull request by VickyStash**
### Details

[TS migration] Migrate 'WorkspaceInvite' page to TypeScript

### Fixed Issues

$ https://github.com/Expensify/App/issues/25207
PROPOSAL: N/A

### Tests

- [x] Verify that no errors appear in the JS console

1. Open the app and login
2. Go to settings -> workspaces -> select any workspace -> members -> invite
3. Verify that user selection/deselection works as expected(toggled)
4. Verify that searching works as expected. Selected options should remain when search query is changed
5. Choose members to invite -> press Next
6. The invite message page should be shown with the chosen members and the invite message.
7. Verify the input works as expected, update the message
8. Press save, make sure members were added to the workspace

### Offline tests

N/A

### QA Steps

- [x] Verify that no errors appear in the JS console

1. Open the app and login
2. Go to settings -> workspaces -> select any workspace -> members -> invite
3. Verify that user selection/deselection works as expected(toggled)
4. Verify that searching works as expected. Selected options should remain when search query is changed
5. Choose members to invite -> press Next
6. The invite message page should be shown with the chosen members and the invite message.
7. Verify the input works as expected, update the message
8. Press save, make sure members were added to the workspace

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/23176449/7d4588ea-c5b7-4c37-8e80-8cedf7c1bf1b

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/23176449/108dde7f-78e2-4cc7-8b55-e6e1f8487ae0

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/23176449/b475f66a-bd0c-48e2-a0a8-e4a4d9a385b2

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/23176449/62926a74-95e4-4c64-a5aa-9039cf5c607f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/23176449/71dac70a-039a-4300-b60e-4fe4be41a2fc

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/23176449/9f0048c9-3066-4273-becf-e19319cd65c7

</details>

### CODE REVIEW

**Review by kubabutkiewicz — APPROVED** [review]

File: src/libs/OptionsListUtils.ts [hunk 1]
```diff
@@ -2003,3 +2005,4 @@ export {
     formatSectionsFromSearchTerm,
     transformedTaxRates,
 };
+export type {MemberForList};
```
Comment (fabioh8010) [inline src/libs/OptionsListUtils.ts:2008]:
```suggestion

export type {MemberForList};
```

File: src/libs/OptionsListUtils.ts [hunk 2]
```diff
@@ -310,9 +310,9 @@ function getPersonalDetailsForAccountIDs(accountIDs: number[] | undefined, perso
 /**
  * Return true if personal details data is ready, i.e. report list options can be created.
  */
-function isPersonalDetailsReady(personalDetails: OnyxEntry<PersonalDetailsList>): boolean {
-    const personalDetailsKeys = Object.keys(personalDetails ?? {});
-    return personalDetailsKeys.some((key) => personalDetails?.[key]?.accountID);
+function isPersonalDetailsReady(personalDetails: OnyxEntry<PersonalDetailsList> | ReportUtils.OptionData[]): boolean {
+    const personalDetailsValues = Array.isArray(personalDetails) ? personalDetails : Object.values(personalDetails ?? {});
+    return personalDetailsValues.some((personalDetail) => personalDetail?.accountID);
 }
```
Comment (fabioh8010) [inline src/libs/OptionsListUtils.ts:316]: Why these changes?

File: src/ONYXKEYS.ts [hunk 3]
```diff
@@ -459,7 +459,8 @@ type OnyxValues = {
     [ONYXKEYS.COLLECTION.POLICY_REPORT_FIELDS]: OnyxTypes.PolicyReportFields;
     [ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_REPORT_FIELDS]: OnyxTypes.RecentlyUsedReportFields;
     [ONYXKEYS.COLLECTION.DEPRECATED_POLICY_MEMBER_LIST]: OnyxTypes.PolicyMembers;
-    [ONYXKEYS.COLLECTION.WORKSPACE_INVITE_MEMBERS_DRAFT]: Record<string, number>;
+    [ONYXKEYS.COLLECTION.WORKSPACE_INVITE_MEMBERS_DRAFT]: OnyxTypes.InvitedEmailsToAccountIDsDraft | undefined;
+    [ONYXKEYS.COLLECTION.WORKSPACE_INVITE_MESSAGE_DRAFT]: string | undefined;
```
Comment (fabioh8010) [inline src/ONYXKEYS.ts:463]: Why are we adding `undefined`?

File: src/pages/workspace/WorkspaceInvitePage.tsx [hunk 4]
```diff
@@ -111,57 +100,69 @@ function WorkspaceInvitePage(props) {
...
-            newSelectedOptions.push(_.has(detailsMap, option.login) ? {...detailsMap[option.login], isSelected: true} : option);
+        selectedOptions.forEach((option) => {
+            newSelectedOptions.push(option.login && option.login in detailsMap ? {...detailsMap[option.login], isSelected: true} : option);
         });
 
         const userToInvite = inviteOptions.userToInvite;
 
         // Only add the user to the invites list if it is valid
-        if (userToInvite) {
+        if (typeof userToInvite?.accountID === 'number') {
```
Comment (fabioh8010) [inline src/pages/workspace/WorkspaceInvitePage.tsx:136]:
```suggestion
        if (userToInvite?.accountID) {
```

Couldn't we just do this?

File: src/pages/workspace/WorkspaceInvitePage.tsx [hunk 5]
```diff
@@ -111,57 +100,69 @@ function WorkspaceInvitePage(props) {
...
+            }
             newPersonalDetailsDict[details.accountID] = details;
         });
 
         // Add all selected options to the new dict
-        _.each(newSelectedOptions, (option) => {
+        newSelectedOptions.forEach((option) => {
+            if (typeof option.accountID !== 'number') {
+                return;
+            }
```
Comment (fabioh8010) [inline src/pages/workspace/WorkspaceInvitePage.tsx:152]: Same

File: src/types/onyx/InvitedEmailsToAccountIDsDraft.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+type InvitedEmailsToAccountIDsDraft = Record<string, number>;
```
Comment (fabioh8010) [inline src/types/onyx/InvitedEmailsToAccountIDsDraft.ts:1]: You can use this type in `src/libs/actions/Policy.ts` too, there are a couple of places there that can use this I think.

File: src/pages/workspace/WorkspaceInvitePage.tsx [hunk 7]
```diff
@@ -219,57 +220,54 @@ function WorkspaceInvitePage(props) {
...
     const inviteUser = () => {
         if (!validate()) {
             return;
         }
 
-        const invitedEmailsToAccountIDs = {};
-        _.each(selectedOptions, (option) => {
-            const login = option.login || '';
-            const accountID = lodashGet(option, 'accountID', '');
+        const invitedEmailsToAccountIDs: Record<string, number> = {};
```
Comment (fabioh8010) [inline src/pages/workspace/WorkspaceInvitePage.tsx:253]:
```suggestion
        const invitedEmailsToAccountIDs: InvitedEmailsToAccountIDsDraft = {};
```

File: src/ONYXKEYS.ts (hunk 3 above)
Comment (VickyStash) [reply src/ONYXKEYS.ts:463]: Please, see the [slack message](https://callstack-hq.slack.com/archives/C05P72RDZD2/p1706195580583269)

File: src/pages/workspace/WorkspaceInvitePage.tsx (hunk 4 above)
Comment (VickyStash) [reply src/pages/workspace/WorkspaceInvitePage.tsx:136]: In theory it's possible `accountID` will be 0 and the check won't work as expected

**Review by fabioh8010 — APPROVED** [review]

File: src/components/SelectionList/types.ts [hunk 8]
```diff
@@ -55,7 +55,7 @@ type User = {
     login?: string;
 
     /** Element to show on the right side of the item */
-    rightElement?: ReactElement;
+    rightElement?: ReactElement | null;
```
Comment (blazejkustra) [inline src/components/SelectionList/types.ts:58]: How about `ReactNode` here?

**Review by blazejkustra — APPROVED** [review]

**Review by hoangzinh — APPROVED** [review]
LGTM

**Review by blimpich — DISMISSED** [review]
LGTM, though we need to resolve the merge conflicts. @VickyStash can you resolve those? Sorry no internal engineer got assigned to this for awhile.

**Comment by VickyStash** [comment]
@blimpich Conflicts are resolved 👌

**Comment by blimpich** [comment]
Hmm, the GitHub UI is still telling me that there are conflicts that need to be resolved

Screenshot 2024-02-09 at 9 13 36 AM

**Comment by VickyStash** [comment]
@blimpich resolved one more time 👌

**Comment by VickyStash** [comment]
Note: I'll be OOO Feb 10 - Feb 18 🌴

**Review by blimpich — APPROVED** [review]
:shipit:

File: src/pages/workspace/WorkspaceInviteMessagePage.tsx [hunk 9]
```diff
@@ -1,139 +1,119 @@
...
-    const [welcomeNote, setWelcomeNote] = useState();
+    const [welcomeNote, setWelcomeNote] = useState<string>();
 
     const {inputCallbackRef} = useAutoFocusInput();
 
     const getDefaultWelcomeNote = () =>
-        props.workspaceInviteMessageDraft ||
+        // workspaceInviteMessageDraft can be an empty string
+        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+        workspaceInviteMessageDraft ||
```
[the remark this answers is not in the document]
Comment (VickyStash) [reply src/pages/workspace/WorkspaceInviteMessagePage.tsx:65]: I think it was to keep the logic consistent with the way it was before the TS migration.
As you can see `props.workspaceInviteMessageDraft ||` was used, and since `workspaceInviteMessageDraft` can be the empty string the change to `??` would affect the logic before the TS migration, we wanted to escape any side updates when it's changes the logic.
